com.unity.multiplayer.mlapi/Editor/CodeGen/NetworkBehaviourILPP.cs

```diff
@@ -578,8 +585,16 @@ private void ProcessNetworkBehaviour(TypeDefinition typeDefinition)
                     instructions.Add(processor.Create(OpCodes.Ldftn, method));
                     instructions.Add(processor.Create(OpCodes.Newobj, m_NetworkHandlerDelegateCtor_MethodRef));
                     instructions.Add(processor.Create(OpCodes.Call, m_NetworkManager_ntable_Add_MethodRef));
+
+                    // NetworkManager.__rpc_name_table.Add(HandlerHash, RpcMethodName);
+                    instructions.Add(processor.Create(OpCodes.Ldsfld, m_NetworkManager_rpc_name_table_FieldRef));
+                    instructions.Add(processor.Create(OpCodes.Ldc_I4, unchecked((int)hash)));
+                    instructions.Add(processor.Create(OpCodes.Ldstr, method.Name.Substring(0, method.Name.Length-"__nhandler".Length)));
+                    instructions.Add(processor.Create(OpCodes.Call, m_NetworkManager_rpc_name_table_Add_MethodRef));
```

Comment:
having another thought on this, I came back to the idea of kicking this (and the static `__rpc_name_table` field in `NetworkManager` class) out from non-editor && non-development builds, in other terms, these stuff should be there in the editor and/or in the development builds only.

if `#if UNITY_EDITOR || DEVELOPMENT_BUILD` macro wouldn't work here, using [`Debug.isDebugBuild`](https://docs.unity3d.com/ScriptReference/Debug-isDebugBuild.html) and [`Application.isEditor`](https://docs.unity3d.com/ScriptReference/Application-isEditor.html) APIs at ILPP runtime would help.

what do you think @Rosme ?

Reply:
We're doing to have some tooling for non-development builds (realtime stats monitor, etc)

Reply:
We can probably include it only by defining only if the conditional dependency of the tooling library is present. If a dev wanted to 100% strip out tooling, they can just remove the lib, but we expect some of them to want this functionality even in non-development builds.

Reply:
would non-dev builds need RPC string names too?

Reply:
@becksebenius-unity

Reply:
We don't have any tools designed that need the finer details of RPC string names yet, so the idea of reporting metrics in non-debug scenarios is more of a general idea and not specific to this metric. That said, even if we don't have a use case for this metric right now, I'm not seeing the harm of including it in release builds - is that something you're advocating for, or just making suggestions for how we'd do it if we were to go down that path?

Reply:
the harm is `Dictionary.Add()` calls in static constructors and a block of managed memory allocated just for RPC name strings. they are not free, we might argue if they're small enough to be no problems but if we don't really have an immediate use-case for them, I'd suggest stripping them out from the non-dev builds. we could put them back (maybe differently and more efficiently) when we really need them (I'm thinking Unity.Collections.FixedString stuff).

Reply:
I'm not seeing the performance concern here. Static constructors are called once-ever for a given class. Dictionary.Add isn't an expensive operation - it's O(1). The memory cost is a single string name for each RPC defined in code - and in fact because this is generated from ILPP, they should be interned by the compiler and not allocated at runtime (even in the constructor). The interned string may even already exist to enable method reflection. I'm not convinced that the code complexity of conditionally running this code is worth it to remove the extremely small runtime impact caused by building a small dictionary.

Reply:
@becksebenius-unity interning can't happen automatically  here because we modify the dll post compile which means string literals injects aren't going to get interned since the compiler is out of it at this point and they won't get added to the intern pool (this is one of the reasons we should get to source code generators at some point) so that means each string is a ldstr opcode that is not interned which means its going to load a copy of the string when it does the add and thus we garbage is generated and a GC will happen for the temp string and further references to the string will result in copies since the interning is not happening. 

Now we could do a manual intern with string.intern which would ensure that, but I have just further issue with your statement, right now we are not using these names and they are not free, they grow unbounded in that the amount of them is N where N is the number of unique RPCs in an application. They are eating up memory that is not being used and I have no way of freeing or clearing that memory and I can assure you any dev is going to say ask why do I have this giant dictionary of strings that does nothing and I don't need 90% of the time. 

To add to that C# static constructors are lazy which  means they do not get called until the type is used or instantiated, which  means this hit does potentially come at any time during the lifetime of the game/application. So when you add this together its not free by any means especially with the lack of interning on the string. 

So I would at least like to ensure that this code doesn't run when the user doesn't want it to run and especially given it has no value at the moment and I can't see devs wanting this overhead especially on servers and mobile devices in production I am just not seeing what we should not either generate this code unless you are in the editor or a dev build or check at runtime.

Reply:
Thanks for correcting my assumption about interning, it makes sense that this happens in IL and therefore won't automatically intern during weaving. Although unless I'm misunderstanding, I think your statement that further references to these strings will result in copies, these will strings would only be allocated once and just the pointer will be passed around.

It sounds like you guys feel strongly about the performance cost here. This smells a bit of preoptimization (I still think the cost is being overstated) but I'm not going to push on this if there is consensus. This will mean that we will be inconsistently applying debug-only rules to the collection of metrics and that adds complexity, although admittedly it is extremely unlikely that we would ever have a use for this data in release builds. We'll need to figure out a good way to document which metrics are release-accessible and which aren't so that this doesn't cause confusion down the road

Reply:
So I don't disagree with you that this is a micro-optimization; but that is not even my primary concern my main issue is actually consistent design principles that we apply to the code base. One of those principles should be that we are ensuring that code that is executing is actually adding value under all circumstances by which the code should be executing. 

This code doesn't pass that test, it executes even when the user may not be actually benefiting from or using it, which to me says the code is optional and thus should not be imposed on the user but in fact truly be optional. 

If we just allow it to be there for the sake of being there regardless if its being used or not in execution that to me sets a bad precedent for future code and to me it says to our users that we don't care about your resources and we are just going to waste cycles/memory and give you no options to disable the functionality.

Reply:
Sorry for the late response. I have read everything. In my opinion, as of now, there is no reason to have those strings in non debug build. I would rather have them removed from the release build, and if needed, find a solution to put them back in an efficient way if we need them in release build. While it may be a micro-optimization, as @wackoisgod explained, it's a matter of consistences and there is still a small cost. Might as well limit that for now.